src/fp.rs

```diff
@@ -0,0 +1,295 @@
+// Copyright (c) 2021 The Authors
+// SPDX-License-Identifier: MPL-2.0
+
+use rand::prelude::*;
+use rand::Rng;
+
+// Parameters of a finite field with prime p < 2^126. Field elements are represented as u128's.
+#[derive(Debug)]
+pub struct Fp {
+    pub p: u128,  // the prime modulus p
+    pub p2: u128, // p * 2
+    pub mu: u64,  // mu = -p^(-1) mod 2^64
+    pub r2: u128, // r2 = (2^128)^2 mod p
+}
+
+impl Fp {
+    pub fn add(&self, x: u128, y: u128) -> u128 {
+        let (z, carry) = x.wrapping_add(y).overflowing_sub(self.p2);
+        let m = 0u128.wrapping_sub(carry as u128);
+        z.wrapping_add(m & self.p2)
+    }
+
+    pub fn sub(&self, x: u128, y: u128) -> u128 {
+        let (z, carry) = x.overflowing_sub(y);
+        let m = 0u128.wrapping_sub(carry as u128);
+        z.wrapping_add(m & self.p2)
+    }
+
+    pub fn mul(&self, x: u128, y: u128) -> u128 {
```

Comment:
It'd be good for all the `pub` functions on `Fp` to have doccomments, especially `mul` which merits some more explanation. At least a link to a reference describing how it works.

Reply:
I added docucomments to all of the methods. @armfazh, can you provide a pointer that describes Montgomery multiplication? I found https://en.wikipedia.org/wiki/Montgomery_modular_multiplication ... not sure of the quality of the description there.

Reply:
Montgomery REDC
 https://www.ams.org/journals/mcom/1985-44-170/S0025-5718-1985-0777282-X/S0025-5718-1985-0777282-X.pdf

Reply:
Added the reference.